src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -774,25 +767,45 @@ partMonad g = asum $ map (runDfs g) $ AM.vertexList g
 -- 'Data.Either.isRight' (detectParts ('Algebra.Graph.AdjacencyMap.star' x ys))                       == not (elem x ys)
 -- 'Data.Either.isRight' (detectParts ('fromBipartite' ('toBipartite' x)))   == True
 -- @
-detectParts :: Ord a => AM.AdjacencyMap a -> Either (OddCycle a) (AdjacencyMap a a)
-detectParts g = let s = AM.symmetricClosure g
-                 in case runState (runMaybeT $ partMonad s) Map.empty of
-                         (Nothing, m) -> Right $ build m g
-                         (Just c,  _) -> Left  $ oddCycle c
+detectParts :: forall a. Ord a => AM.AdjacencyMap a -> Either (OddCycle a) (AdjacencyMap a a)
+detectParts x = case runState (runMaybeT $ dfs) Map.empty of
+                     (Nothing, m) -> Right $ toBipartiteWith (toEither m) g
```

Comment:
It looks like this function may be doing more than it needs to: it could return `a -> Either b c` as a partition function instead of a returning an actual bipartite graph. The user could then use `toBipartiteWith` if need be.

We could also add a type synonym `type Partition a b c = a -> Either b c`.

Reply:
For the history sake, I'm copying the discussion here.

This seemed a really nice idea, but there are several problems with it, the main being that it's not clear what should be returned by `Partition` in case it's given a vertex that is not present in the graph. Some workaround could be suggested, for example,

```Haskell
type Partition a b c = a -> Maybe (Either b c)
```

But there are still problems with this, particularly that `Nothing` could be returned for a present vertex and that this one is anyway too complicated for use.

We'll keep the current implementation with `Either (OddCycle a) (AdjacencyMap a a)`.